homeassistant/components/tile/__init__.py

```diff
@@ -1,137 +1,93 @@
-    CoordinatorEntity,
-    DataUpdateCoordinator,
-    UpdateFailed,
-)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import DATA_COORDINATOR, DOMAIN, LOGGER
+from .const import DATA_COORDINATOR, DATA_TILE, DOMAIN, LOGGER
 
 PLATFORMS = ["device_tracker"]
 DEVICE_TYPES = ["PHONE", "TILE"]
 
-DEFAULT_ATTRIBUTION = "Data provided by Tile"
-DEFAULT_ICON = "mdi:view-grid"
 DEFAULT_UPDATE_INTERVAL = timedelta(minutes=2)
 
 CONF_SHOW_INACTIVE = "show_inactive"
 
 
 async def async_setup(hass, config):
     """Set up the Tile component."""
-    hass.data[DOMAIN] = {DATA_COORDINATOR: {}}
-
+    hass.data[DOMAIN] = {DATA_COORDINATOR: {}, DATA_TILE: {}}
     return True
 
 
-async def async_setup_entry(hass, config_entry):
+async def async_setup_entry(hass, entry):
     """Set up Tile as config entry."""
+    hass.data[DOMAIN][DATA_COORDINATOR][entry.entry_id] = {}
+    hass.data[DOMAIN][DATA_TILE][entry.entry_id] = {}
+
     websession = aiohttp_client.async_get_clientsession(hass)
 
-    client = await async_login(
-        config_entry.data[CONF_USERNAME],
-        config_entry.data[CONF_PASSWORD],
-        session=websession,
-    )
+    try:
+        client = await async_login(
+            entry.data[CONF_USERNAME],
+            entry.data[CONF_PASSWORD],
+            session=websession,
+        )
+        hass.data[DOMAIN][DATA_TILE][entry.entry_id] = await client.async_get_tiles()
+    except TileError as err:
+        LOGGER.error("Error during integration setup: %s", err)
+        return False
 
-    async def async_update_data():
-        """Get new data from the API."""
+    async def async_update_tile(tile):
+        """Update the Tile."""
         try:
-            return await client.tiles.all()
+            return await tile.async_update()
         except SessionExpiredError:
             LOGGER.info("Tile session expired; creating a new one")
             await client.async_init()
         except TileError as err:
             raise UpdateFailed(f"Error while retrieving data: {err}") from err
 
-    coordinator = DataUpdateCoordinator(
-        hass,
-        LOGGER,
-        name=config_entry.title,
-        update_interval=DEFAULT_UPDATE_INTERVAL,
-        update_method=async_update_data,
-    )
+    coordinator_init_tasks = []
+    for tile_uuid, tile in hass.data[DOMAIN][DATA_TILE][entry.entry_id].items():
+        coordinator = hass.data[DOMAIN][DATA_COORDINATOR][entry.entry_id][
+            tile_uuid
+        ] = DataUpdateCoordinator(
+            hass,
+            LOGGER,
+            name=tile.name,
+            update_interval=DEFAULT_UPDATE_INTERVAL,
+            update_method=partial(async_update_tile, tile),
+        )
+        coordinator_init_tasks.append(coordinator.async_refresh())
 
-    await coordinator.async_refresh()
-    hass.data[DOMAIN][DATA_COORDINATOR][config_entry.entry_id] = coordinator
+    await asyncio.gather(*coordinator_init_tasks)
```

Comment:
This could end up making a lot of requests to the tile server. We have a helper that can limit the number of parallel requests https://github.com/home-assistant/core/blob/dev/homeassistant/util/async_.py#L126